### PR TITLE
Exercises

### PR DESCRIPTION
**Pull request by jburel**
Add exercise to show to crop an image and import the cropped image
cc @pwalczysko

### CODE REVIEW

File: practical/groovy/crop_rectangle_from_image.groovy [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * ------------------------------------------------------------------------------
+ */
+
+/*
+ * This Groovy script uses ImageJ to crop an image.
+ * The cropped image is saved locally as OME-TIFF using Bio-Formats Exporter.
+ * Upload the OME-TIFF to OMERO.
```
Comment (pwalczysko) [inline practical/groovy/crop_rectangle_from_image.groovy]: Maybe ``Upload the OME....`` -> ``The OME-TIFF is then uploaded to OMERO.``
?

Comment (pwalczysko) [reply practical/groovy/crop_rectangle_from_image.groovy]: Maybe even better: ``The script then uploads the OME-TIFF to OMERO, to a new dataset named "Cropped".``

File: practical/groovy/crop_rectangle_from_image.groovy [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+println "connecting..."
+gateway = connect_to_omero()
+
+println "opening Image..."
+// Open the Image using Bio-Formats
+open_image_plus(HOST, USERNAME, PASSWORD, PORT, group_id, image_id)
+
+// Crop the image
+println "cropping..."
+IJ.makeRectangle(0, 0, 50, 50)
```
Comment (pwalczysko) [inline practical/groovy/crop_rectangle_from_image.groovy]: very small rectangle - it is actually quite hard to believe this is a crop of the image I have passed into the script - maybe we could go for at least twice as large one ?

File: practical/groovy/background_subtract_batch.groovy [hunk 3]
```diff
@@ -187,7 +187,10 @@ image_ids.each() { image_id ->
 
     // Save modified image as OME-TIFF using Bio-Formats
```
Comment (pwalczysko) [inline practical/groovy/background_subtract_batch.groovy]: Not sure here - the exercise 2 in the workflow is asking me to upload a Map Annotation . What do I need a bckgr subtraction for then ?

Comment (jburel) [reply practical/groovy/background_subtract_batch.groovy]: background script won't be included I just noticed that it was not created a file in the tmp dir
So I updated it
I can do that in another PR but the time is not on our side

**Comment by pwalczysko** [comment]
Checked the last commit, +1 for the bigger crop and the text.

File: practical/groovy/background_subtract_batch.groovy (hunk 3 above)
Comment (pwalczysko) [reply practical/groovy/background_subtract_batch.groovy]: No, no need for extra work, I was just confused, leave as it is please

**Comment by jburel** [comment]
@pwalczysko example 2 added

File: practical/groovy/analyse_image_map_annotation.groovy [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    gateway = new Gateway(simpleLogger)
+    gateway.connect(credentials)
+    return gateway
+
+}
+
+def open_image_plus(HOST, USERNAME, PASSWORD, PORT, group_id, image_id) {
+    "Open the image using the Bio-Formats Importer"
+
+    StringBuffer options = new StringBuffer()
```
Comment (mtbc) [inline practical/groovy/analyse_image_map_annotation.groovy]: Use `StringBuilder` if you don't need thread-safety.

Comment (jburel) [reply practical/groovy/analyse_image_map_annotation.groovy]: I will keep it as it is for now since I do not have time to review all the other scripts.

Comment (mtbc) [reply practical/groovy/analyse_image_map_annotation.groovy]: Okay.

Comment (jburel) [reply practical/groovy/analyse_image_map_annotation.groovy]: Saying that I will fix those 2 and open another PR
the power of cut & paste :-)

**Comment by pwalczysko** [comment]
The https://github.com/ome/training-scripts/pull/59/commits/e7246940696aa50dd25291f1f7df38cc26a51173 is working - the cropping script still works as expected.

File: practical/groovy/analyse_image_map_annotation.groovy [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * ------------------------------------------------------------------------------
+ */
+
+/*
+ * This Groovy script uses ImageJ to crop an image.
```
Comment (pwalczysko) [inline practical/groovy/analyse_image_map_annotation.groovy]: The description needs to be updated, so that it refers to the map annotations, not crop.

**Comment by pwalczysko** [comment]
screen shot 2019-02-28 at 10 14 29

I would find it helpful if the last message in the output window would be "Done", was still waiting even after the script finished.


Otherwise, works as expected.
The only thing is that somebody might be surprised that, although the Map Ann says "11 ROIs", there might be another number of ROIs on that image, as the script does not save ROIs. But I guess we can explain that, it can be instructive.


One small issue with "Done", otherwise all good here (except I need "white" for my version of Fiji)

File: practical/groovy/analyse_image_map_annotation.groovy [hunk 6]
```diff
@@ -20,9 +20,10 @@
  */
 
 /*
- * This Groovy script uses ImageJ to crop an image.
- * The cropped image is saved locally as OME-TIFF using Bio-Formats Exporter.
- * Upload the OME-TIFF to OMERO.
+ * This Groovy script uses ImageJ to analyse an Image.
+ * The number of generated ROIs is then adding alongside other information
```
Comment (pwalczysko) [inline practical/groovy/analyse_image_map_annotation.groovy]: adding -? added

**Comment by pwalczysko** [comment]
One small comment on the last (description text) commit https://github.com/ome/training-scripts/pull/59/commits/b45ca4934fdc3d5b660f68d581bef7820fccbeef#r261128957

**Comment by pwalczysko** [comment]
Map Annotation script -> works as expected.

**Comment by pwalczysko** [comment]
Crop script -> works as expected. Ready to merge FMPOV

**Pull request merged**